crates/task-impls/src/consensus.rs

```diff
@@ -581,245 +681,192 @@ where
+                } else {
+                    consensus
+                        .saved_leaves
+                        .get(&justify_qc.leaf_commitment())
+                        .cloned()
+                };
+
+                //
+                // Justify qc's leaf commitment is not the same as the parent's leaf commitment, but it should be (in this case)
+                let Some(parent) = parent else {
+                    // If no parent then just update our state map and return.  We will not vote.
+                    error!(
+                        "Proposal's parent missing from storage with commitment: {:?}",
+                        justify_qc.leaf_commitment()
+                    );
+                    let leaf = SequencingLeaf {
+                        view_number: view,
+                        height: proposal.data.height,
+                        justify_qc: justify_qc.clone(),
+                        parent_commitment: justify_qc.leaf_commitment(),
+                        deltas: Right(proposal.data.block_commitment),
+                        rejected: Vec::new(),
+                        timestamp: time::OffsetDateTime::now_utc().unix_timestamp_nanos(),
+                        proposer_id: sender.to_bytes(),
+                    };
+
+                    let mut consensus = RwLockUpgradableReadGuard::upgrade(consensus).await;
+                    consensus.state_map.insert(
+                        view,
+                        View {
+                            view_inner: ViewInner::Leaf {
+                                leaf: leaf.commit(),
+                            },
+                        },
+                    );
+                    consensus.saved_leaves.insert(leaf.commit(), leaf.clone());
 
-                            if invalid {
-                                error!("Invalid justify_qc in proposal {:?}", justify_qc.clone());
-                                return;
-                            }
-                            leaf = SequencingLeaf {
-                                view_number: view,
-                                height: proposal.data.height,
-                                justify_qc: justify_qc.clone(),
-                                parent_commitment: justify_qc.leaf_commitment(),
-                                deltas: Right(proposal.data.block_commitment),
-                                rejected: Vec::new(),
-                                timestamp: time::OffsetDateTime::now_utc().unix_timestamp_nanos(),
-                                proposer_id: sender.to_bytes(),
-                            };
-                            if !view_leader_key
-                                .validate(&proposal.signature, leaf.commit().as_ref())
-                            {
-                                error!(?proposal.signature, "Could not verify proposal.");
-                                return;
-                            }
+                    return;
```

Comment:
In the fix branch we were updating the view here to avoid timing out in the view that might have actually been successful.  I'm not 100% sure either way so I'll leave up to you.  If we don't update the view like in this PR, then the view will time out and the network might move on to future views without us and we'll be stuck trying to catch back up.

If we do update the view then we might time out on a future view when everyone else times out because this is a bogus QC that nobody voted on.

Reply:
If after this is deployed and we see one node constantly behind, this is probably the culprit.  On the other hand if we see nodes entering view sync at different times in the current deployment this change would fix it.